File: examples/mcp_client_example.js
#!/usr/bin/env node
/**
 * Example MCP client for the Druids multi-agent system (Node.js version).
 * This script demonstrates how to connect to and interact with the Druids MCP server.
 * 
 * Usage:
 *   node examples/mcp_client_example.js
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');

class DruidsMCPClient {
    constructor(baseUrl = 'http://localhost:3003') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'User-Agent': 'DruidsMCPClient-Node/1.0.0'
        };
    }

    async request(method, path, data = null) {
        return new Promise((resolve, reject) => {
            const url = new URL(path, this.baseUrl);
            const options = {
                hostname: url.hostname,
                port: url.port,
                path: url.pathname + url.search,
                method: method,
                headers: this.headers
            };

            const client = url.protocol === 'https:' ? https : http;
            const req = client.request(options, (res) => {
                let body = '';
                res.on('data', chunk => body += chunk);
                res.on('end', () => {
                    try {
                        const result = JSON.parse(body);
                        if (res.statusCode >= 200 && res.statusCode < 300) {
                            resolve(result);
                        } else {
                            reject(new Error(`HTTP ${res.statusCode}: ${result.error || body}`));
                        }
                    } catch (e) {
                        reject(new Error(`Invalid JSON response: ${body}`));
                    }
                });
            });

            req.on('error', reject);

            if (data) {
                req.write(JSON.stringify(data));
            }
            req.end();
        });
    }

    async healthCheck() {
        return this.request('GET', '/health');
    }

    async getCapabilities() {
        return this.request('GET', '/mcp/capabilities');
    }

    async listTools() {
        const result = await this.request('GET', '/mcp/tools');
        return result.tools;
    }

    async callTool(toolName, args) {
        return this.request('POST', `/mcp/tools/${toolName}/call`, { arguments: args });
    }

    async listResources() {
        const result = await this.request('GET', '/mcp/resources');
        return result.resources;
    }

    async getResource(resourceId) {
        return this.request('GET', `/mcp/resources/${resourceId}`);
    }

    async listPrompts() {
        const result = await this.request('GET', '/mcp/prompts');
        return result.prompts;
    }

    async executePrompt(promptName, args) {
        return this.request('POST', `/mcp/prompts/${promptName}`, { arguments: args });
    }
}

async function main() {
    const client = new DruidsMCPClient();

    try {
        // Check server health
        console.log('🔍 Checking server health...');
        const health = await client.healthCheck();
        console.log(`✅ Server is healthy: ${health.status}`);
        console.log();

        // Get capabilities
        console.log('🔧 Getting server capabilities...');
        const capabilities = await client.getCapabilities();
        console.log(`📋 Server: ${capabilities.implementation.name} v${capabilities.implementation.version}`);
        console.log();

        // List available tools
        console.log('🛠️ Available tools:');
        const tools = await client.listTools();
        tools.forEach(tool => {
            console.log(`  • ${tool.name}: ${tool.description}`);
        });
        console.log();

        // Example: Create a realm
        console.log('🏞️ Creating a test realm...');
        const realmResult = await client.callTool('realm_create', {
            name: 'Node.js Client Test Realm',
            type: 'mountain',
            description: 'A test realm created via Node.js MCP client'
        });
        console.log(`✅ Realm creation result: ${realmResult.content[0].text}`);
        console.log();

        // Example: Create an agent
        console.log('🧙‍♂️ Creating a test agent...');
        const agentResult = await client.callTool('agent_create', {
            name: 'Node.js Test Elemental',
            type: 'elemental',
            description: 'A test elemental created via Node.js MCP client',
            realm: 'Node.js Client Test Realm'
        });
        console.log(`✅ Agent creation result: ${agentResult.content[0].text}`);
        console.log();

        // Example: Query knowledge
        console.log('🔍 Querying knowledge...');
        const knowledgeResult = await client.callTool('knowledge_query', {
            namespace: 'test-namespace',
            query: 'multi-agent systems',
            limit: 5
        });
        console.log(`✅ Knowledge query result: ${knowledgeResult.content[0].text}`);
        console.log();

        // Example: Generate scenario plan
        console.log('📜 Generating scenario plan...');
        const scenarioPlan = await client.executePrompt('scenario_plan', {
            objective: 'Coordinate forest preservation efforts',
            agents: ['Node.js Test Elemental', 'Forest Guardian Druid', 'Gaia Monitor']
        });
        console.log('📋 Generated scenario plan:');
        console.log(scenarioPlan.messages[0].content.text);
        console.log();

        // List available resources
        console.log('📚 Available resources:');
        const resources = await client.listResources();
        resources.forEach(resource => {
            console.log(`  • ${resource.uri}: ${resource.description}`);
        });
        console.log();

        // Get agents resource
        console.log('👥 Getting agents resource...');
        const agentsResource = await client.getResource('agents');
        console.log(`📊 Agents data: ${agentsResource.contents[0].text}`);
        console.log();

        console.log('🎉 All MCP operations completed successfully!');

    } catch (error) {
        console.error(`❌ Error: ${error.message}`);
        console.log('💡 Make sure the Druids system is running: ./scripts/dev.sh start');
    }
}

// Run the example
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { DruidsMCPClient };

File: examples/mcp_client_example.py
#!/usr/bin/env python3
"""
Example MCP client for the Druids multi-agent system.
This script demonstrates how to connect to and interact with the Druids MCP server.

Usage:
    python examples/mcp_client_example.py
"""

import json
import requests
from typing import Dict, Any, List

class DruidsMCPClient:
    """Simple MCP client for the Druids system."""
    
    def __init__(self, base_url: str = "http://localhost:3003"):
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'User-Agent': 'DruidsMCPClient/1.0.0'
        })
    
    def health_check(self) -> Dict[str, Any]:
        """Check if the MCP server is healthy."""
        response = self.session.get(f"{self.base_url}/health")
        response.raise_for_status()
        return response.json()
    
    def get_capabilities(self) -> Dict[str, Any]:
        """Get server capabilities."""
        response = self.session.get(f"{self.base_url}/mcp/capabilities")
        response.raise_for_status()
        return response.json()
    
    def list_tools(self) -> List[Dict[str, Any]]:
        """List available tools."""
        response = self.session.get(f"{self.base_url}/mcp/tools")
        response.raise_for_status()
        return response.json()['tools']
    
    def call_tool(self, tool_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a tool with given arguments."""
        payload = {"arguments": arguments}
        response = self.session.post(
            f"{self.base_url}/mcp/tools/{tool_name}/call",
            json=payload
        )
        response.raise_for_status()
        return response.json()
    
    def list_resources(self) -> List[Dict[str, Any]]:
        """List available resources."""
        response = self.session.get(f"{self.base_url}/mcp/resources")
        response.raise_for_status()
        return response.json()['resources']
    
    def get_resource(self, resource_id: str) -> Dict[str, Any]:
        """Get a specific resource."""
        response = self.session.get(f"{self.base_url}/mcp/resources/{resource_id}")
        response.raise_for_status()
        return response.json()
    
    def list_prompts(self) -> List[Dict[str, Any]]:
        """List available prompts."""
        response = self.session.get(f"{self.base_url}/mcp/prompts")
        response.raise_for_status()
        return response.json()['prompts']
    
    def execute_prompt(self, prompt_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a prompt with given arguments."""
        payload = {"arguments": arguments}
        response = self.session.post(
            f"{self.base_url}/mcp/prompts/{prompt_name}",
            json=payload
        )
        response.raise_for_status()
        return response.json()

def main():
    """Example usage of the Druids MCP client."""
    client = DruidsMCPClient()
    
    try:
        # Check server health
        print("🔍 Checking server health...")
        health = client.health_check()
        print(f"✅ Server is healthy: {health['status']}")
        print()
        
        # Get capabilities
        print("🔧 Getting server capabilities...")
        capabilities = client.get_capabilities()
        print(f"📋 Server: {capabilities['implementation']['name']} v{capabilities['implementation']['version']}")
        print()
        
        # List available tools
        print("🛠️ Available tools:")
        tools = client.list_tools()
        for tool in tools:
            print(f"  • {tool['name']}: {tool['description']}")
        print()
        
        # Example: Create a realm
        print("🏞️ Creating a test realm...")
        realm_result = client.call_tool("realm_create", {
            "name": "Python Client Test Realm",
            "type": "forest",
            "description": "A test realm created via Python MCP client"
        })
        print(f"✅ Realm creation result: {realm_result['content'][0]['text']}")
        print()
        
        # Example: Create an agent
        print("🧙‍♂️ Creating a test agent...")
        agent_result = client.call_tool("agent_create", {
            "name": "Python Test Druid",
            "type": "druid",
            "description": "A test druid created via Python MCP client",
            "realm": "Python Client Test Realm"
        })
        print(f"✅ Agent creation result: {agent_result['content'][0]['text']}")
        print()
        
        # Example: Execute a scenario
        print("🎭 Executing a test scenario...")
        scenario_result = client.call_tool("scenario_execute", {
            "name": "Python Client Test Scenario",
            "description": "Multi-agent collaboration test from Python client",
            "agents": ["Python Test Druid", "System Agent"]
        })
        print(f"✅ Scenario execution result: {scenario_result['content'][0]['text']}")
        print()
        
        # Example: Generate agent instructions
        print("📜 Generating agent instructions...")
        instructions = client.execute_prompt("agent_instruction", {
            "agent_type": "druid",
            "role": "forest guardian and ecosystem monitor"
        })
        print("📋 Generated instructions:")
        print(instructions['messages'][0]['content']['text'])
        print()
        
        # List available resources
        print("📚 Available resources:")
        resources = client.list_resources()
        for resource in resources:
            print(f"  • {resource['uri']}: {resource['description']}")
        print()
        
        print("🎉 All MCP operations completed successfully!")
        
    except requests.RequestException as e:
        print(f"❌ Error connecting to MCP server: {e}")
        print("💡 Make sure the Druids system is running: ./scripts/dev.sh start")
    except Exception as e:
        print(f"❌ Unexpected error: {e}")

if __name__ == "__main__":
    main()